### ISSUE TITLE
process for publishing PhET-iO dev versions

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/chipper#440]
Discussion about the process of deploying dev versions for PhET-iO was sparked by https://github.com/phetsims/beers-law-lab/issues/153.  There is currently no process.

Decisions:
• what does the version identifier look like?
• is the major.minor.maintenance part of the version identifier shared with non-PhET-iO dev versions?
• where are PhET-iO dev versions deployed?
• how are PhET-iO dev versions kept private?

**Comment by pixelzoom** [comment]
@samreid If there is no process for publishing PhET-iO dev versions, how have you been providing dev versions to clients?

**Comment by pixelzoom** [comment]
Related to https://github.com/phetsims/together/issues/331

**Comment by pixelzoom** [comment]
https://github.com/phetsims/together/issues/331 is a duplicate of this issue, and seems misplaced (in together, should be in chipper).   I'm going to close that issue.  Here's the one relevant comment:

In https://github.com/phetsims/together/issues/331#issue-130914970, @samreid said:
 > Whenever publishing a dev version of a phet-io sim, also publish a dev version of the non-phet-io sim
Also, use .htaccess to password protect the phet-io file.

**Comment by pixelzoom** [comment]
Other relevant comments from https://github.com/phetsims/beers-law-lab/issues/153:

@jonathanolson suggested:
> don't create a separate directory, just use .htaccess to make restrict access to those.

@pixelzoom suggested:
> <FilesMatch> in .htpassword

**Comment by samreid** [comment]
> @samreid If there is no process for publishing PhET-iO dev versions, how have you been providing dev versions to clients?

Custom publications on spot at html/phet-io

**Comment by pixelzoom** [comment]
Does "custom publications" involve manually copying files around, or is there chipper support?

**Comment by samreid** [comment]
> Does "custom publications" involve manually copying files around, or is there chipper support?

I had changed a line or two in deploy-dev.sh to publish these to a different directory.

**Comment by pixelzoom** [comment]
deploy-dev.sh no longer exists. I'm guessing that means you've changed your working copy, but not pushed.

**Comment by pixelzoom** [comment]
Notes from 2/4/16 phet-io meeting:

SR: published versions so far have been through no RC test process
AP: move dev to phet-test, or somewhere that requires VPN
JO: has a preference for a cookie-based password solution, so doesn’t have to log in via a VPN
KP: have a phetio subdir (eg dev/html/concentration/1.1.0-dev.3/phetio/), so that someone with a password could get to it
ran out of time, discussion to be continued

**Comment by samreid** [comment]
@pixelzoom suggested deploying wrappers with each phet-io simulation, which seems nice
@samreid pointed out that we need to make sure that one wrapper works properly with a suite of simulations (and easily test that it is so)

Perhaps we should do both?

**Comment by pixelzoom** [comment]
My recommendation is to publish each sim independently, with all the resources related to it in its deployment directory (including all documentation, and *especially* any source files required to use it, like SimIFrameAPI.js).  The organization of together/phet-io-site/ that you demonstrated in today's meeting is not scalable, and makes it difficult to determine which sim versions go with which other published materials.  And it encourages the client (3rd-party or PhET) to attempt to reuse SimIFrameAPI.js for multiple sims.  Sims evolve independently, and that approach will quickly run into problem (which will be difficult to troubleshoot).  I think we want to encourage clients to use 1 version of provide source code with 1 sim.   This applies to dev and production versions.

**Comment by samreid** [comment]
That sounds pretty great, my main questions (which I suspect can be solved) are:

1. When we deliver 5 sims to a client, they will get 5 identical copies of the "Developer's Guide"?  Perhaps this is OK.
2. When they load Beer's Law Lab, they use the Beer's Law Lab SimIFrameClient.js?  What about wrappers that allow the student to choose from several sims?  Would it dynamically select the appropriate SimIFrameClient.js file?

I also realized that the phet-io simulations on figaro were going to be public (not password protected), so that students don't have to enter a password to use the sims.  Perhaps this means it is OK to publish publicly accessible phet-io sims on spot?  We would still restrict access to the SimIFrameClient.js, example wrappers and API files.

**Comment by samreid** [comment]
I would love a process where publishing to dev and prod are identical aside from the path.

**Comment by samreid** [comment]
@kathy-phet said:

> If I follow, it seems like suggesting that publish one to one package of all you need. This sounds fine as long as Api, etc in subdirectory with password. Main io file is ok to have publicly visible.

**Comment by samreid** [comment]
From the Beer's Law Lab issue:

@pixelzoom said:

The problem here is this awful "2 part" hack that was added to getBuildConfig.js:

```js
  // TODO: chipper#270 workaround, part 1 (see part 2 in createHTMLFilename.js)
  if ( buildConfig.brand === 'phet-io' && buildConfig.version.indexOf( '-dev' ) > -1 ) {
    buildConfig.version = buildConfig.version.replace( '-dev', '-phet-io' );
  }
```

and createHTMLFilename.js:

```js
    // TODO: chipper#270 workaround, part 2 (see part 1 in getBuildConfig.js)
    if ( buildConfig.brand === 'phet-io' ) {
      var devVersion = buildConfig.version.replace( '-phet-io', '-dev' );
      string = string.replace( devVersion, buildConfig.version );
    }
```

**Comment by samreid** [comment]
I published concentration-phetio on spot in https://github.com/phetsims/concentration/issues/48, and wanted to point out these issues:
1. the phet-io version and main version have different version numbers due to the workaround noted in the preceding comment
2. the phet-io version is publicly accessible and publicly visible.  This isn't a tested one, so we may want to hide/password protect it anyways, even though we normally won't password protect the phet-io sims.
3. if someone launches the phet-io sim by itself, it will show the splash screen forever.  When running without a wrapper, the sim requires a query parameter phet-io-standalone to finish launching (indicates that is no wrapper initialization phase).

**Comment by pixelzoom** [comment]
Re questions in https://github.com/phetsims/chipper/issues/440#issuecomment-180108072:

> (1) When we deliver 5 sims to a client, they will get 5 identical copies of the "Developer's Guide"? Perhaps this is OK.

Yes. After the initial delivery to a client, those 5 sims will continue to evolve independently, and it's highly unlikely we'll revise all 5 sims every time we re-deliver.  So it's important that each sim be "standalone", and delivered with its own supporting materials.

> (2) When they load Beer's Law Lab, they use the Beer's Law Lab SimIFrameClient.js? What about wrappers that allow the student to choose from several sims? Would it dynamically select the appropriate SimIFrameClient.js file?

We should encourage clients to use each sim version with the SimIFrameClient.js that it was delivered with.  If they try to use one instance of SimIFrameClient.js for multiple sims, they are likely to run into problems as the sims (and PhET-iO) evolve.

If choosing from several sims can only be done by via one instance of SimIFrameClient.js, then that is a design problem.  It essentially requires all of the sims to be on the same "version" of SimIFrameClient.js, and that most certainly will not be a reasonable constraint in practice.

> (3) How will a client identify whether two simulations are compatible with a given wrapper? They will generally want to develop a single wrapper that works across many sims.

If there's not already something in SimIFrameClient.js that addresses this (version handshaking, or other way to confirm compatibility on start up) then that is a rather large design oversight, and should be addressed.  The software should be able to detect incompatibility.

**Comment by samreid** [comment]
Any suggestions how grunt can build both a phet brand and phet-io brand version?

**Comment by samreid** [comment]
I'm going to try a new task that runs both:

```js
      // Do the phet brand build
      grunt.option( 'brand', 'phet' );
      grunt.task.run( 'build' );

      // Do the phet-io brand build
      grunt.option( 'brand', 'phet-io' );
      grunt.task.run( 'build' );
```

I'll need to add a new option that preserves the build directory instead of clearing it.

**Comment by samreid** [comment]
> I'll need to add a new option that preserves the build directory instead of clearing it.

Or another temporary directory...

**Comment by samreid** [comment]
getBuildConfig includes many fields that are computed based on the brand, and it would be tricky/confusing to try to switch out the brand after some things are initialized.  Options: move to a separate process (one grunt invocation per brand) or rewrite the way brands are handled throughout chipper to make it mutable.

**Comment by samreid** [comment]
Running two processes seems like it is working well.

**Comment by pixelzoom** [comment]
@samreid When you complete this work, can you summarize which grunt tasks this affects, how to publish a PhET-iO dev version, where the files get published to, etc?

**Comment by samreid** [comment]
Let me do my best to answer this now while it is on our minds:

1. cd beers-law-lab
2. grunt build-phet-io

From the grunt help:

```
build-phet-io               Full PhET-iO simulation build including PhET       
                            branded version, PhET-iO branded version and       
                            accompanying PhET-iO material  
```

* This didn't significantly affect the semantics/usage/behavior of any pre-existing tasks.
* Dev files would get published with `grunt deploy-dev`
* Production file deploy is still under consideration, perhaps it could be a new task "grunt deploy-phet-io".  But now that we've matched our dev phet-io deploys to our dev phet brand deploys, it might be straightforward to use deploy-prod to deploy phet-io sims to figaro.  The issue is that @kathy-phet previously said she wants to serve the phet-io sims from a different site.  We should discuss this again and make sure it will be worth the additional complexity to the process.

**Comment by pixelzoom** [comment]
> • Dev files would get published with grunt deploy-dev

Do files phet-branded and phetio-branded files live in the same location on spot? 
Is there any access restriction to the *-phetio.html files?

**Comment by samreid** [comment]
The proposed directory structure is:

```
beers-law-lab/
  1.4.0-dev.3/
    dependencies.json
    beers-law-lab_en.html
    beers-law-lab_en-iframe.html  
    phet-io/
      beers-law-lab_en-phetio.html
      dependencies.json
      devguide/
      protected/
        api/
        lib/
        wrappers/
        wrappers.html
```

The directory "protected" would be password protected.  Since students are the ones who will be launching the instrumented simulations, the instrumented simulations will not be password protected.

**Comment by pixelzoom** [comment]
Thanks for the overview, looks good.

**Comment by samreid** [comment]
I'm thinking of moving the devguide under protected/ as well.

**Comment by samreid** [comment]
I published the first phet-io dev deploy here: http://www.colorado.edu/physics/phet/dev/html/faradays-law/1.1.0-dev.3/

We should take a close look to see if any changes should be made.  It will simplify things if we can use the same directory structure for production phet-io deploys.

**Comment by pixelzoom** [comment]
Dev structure looks good to me, assuming it's OK to have devguide publicly available.

**Comment by samreid** [comment]
PhET-iO deploy is established, there are many examples on the phet-io website, and the instructions are noted at https://github.com/phetsims/phet-info/blob/master/deployment-info/sim_deployment.md
